packet/src/pacer.rs

```diff
@@ -0,0 +1,131 @@
+use std::time::{Duration, Instant};
+
+pub struct LeakyBucketPacer {
+    /// Target bitrate in bits per second.
+    target_bitrate: u64,
+    /// The last time we refreshed the budget.
+    last_budget_increase: Option<Instant>,
+    /// The current budget in bytes.
+    budget: usize,
+    /// The current pacing i.e. how frequently we refresh the budget.
+    pacing: Duration,
+}
+
+pub enum Decision {
+    Emit,
+    Hold,
+}
+
+impl LeakyBucketPacer {
+    pub fn set_target_bitrate(&mut self, new_rate: u64) {
+        self.target_bitrate = new_rate;
+    }
+
+    pub fn next_interval(&self) -> Option<Instant> {
+        self.last_budget_increase.map(|l| l + self.pacing)
+    }
+
+    pub fn decide(&mut self, data: &[u8], now: Instant) -> Decision {
+        self.maybe_update_budget(now);
+        dbg!(data.len(), self.budget);
+
+        if data.len() <= self.budget {
+            self.budget -= data.len();
+            Decision::Emit
+        } else {
+            Decision::Hold
+        }
+    }
+
+    fn new(target_bitrate: u64, pacing: Duration) -> Self {
+        Self {
+            target_bitrate,
+            last_budget_increase: None,
+            budget: 0,
+            pacing,
+        }
+    }
+
+    fn maybe_update_budget(&mut self, now: Instant) {
+        let elapsed = self
+            .last_budget_increase
+            .map(|l| now - l)
+            .unwrap_or(self.pacing);
+
+        if elapsed >= self.pacing {
+            let budget = Self::calculate_budget(self.target_bitrate, elapsed);
+            self.budget = budget;
+            self.last_budget_increase = Some(now);
+            debug!(
+                "Budget increase, new budget: {}, elapsed: {:?}",
+                self.budget, elapsed
+            );
+        }
+    }
+
+    fn calculate_budget(bitrate: u64, duration: Duration) -> usize {
+        ((bitrate as f64 / 8000.0_f64) * (duration.as_millis() as f64)) as usize
+    }
+}
+
+impl Decision {
+    pub fn emit(&self) -> bool {
+        matches!(self, Self::Emit)
+    }
+}
```

Comment:
Isn't it easier to call `decide` something like `try_allocate_budget(&self, len: usize) -> bool` and return a bool? This enum. Does it clarify?

Maybe it's more clear that the function is a bit more precise in the name?

Reply:
👍